blogpost/post.md

```diff
@@ -0,0 +1,263 @@
+# Compiling NumPy code into C++ or CUDA via `torch.compile`
+
+Quansight engineers have implemented the ability to trace through NumPy code via
+`torch.compile` for PyTorch 2.1.
+This feature leverages PyTorch's compiler to generate efficient fused
+vectorized code without having to modify your original code. Even more, it
+also allows for executing NumPy functions on CUDA just by running them through
+`torch.compile` under `torch.device("cuda")`!
+
+In this post, we go over how to use this feature and give a few tips and tricks
+to make the most out of it.
+
+
+## Compiling NumPy code into Parallel C++
+
+We will take as our running example the iteration step in a K-Means algorithm
+presented in this [NumPy book](https://realpython.com/numpy-array-programming/#clustering-algorithms)
+
+```python
+import numpy as np
+
+def get_labels(X, means):
+    return np.argmin(np.linalg.norm(X - means[:, None], axis=2), axis=0)
+```
+
+We create a synthetic dataset with 10M random 2-D points. We can see that,
+given that the means are chosen appropriately, the function returns the correct
+cluster for all of them
+
+```python
+npts = 10_000_000
+X = np.repeat([[5, 5], [10, 10]], [npts, npts], axis=0)
+X = X + np.random.randn(*X.shape)  # 2 distinct "blobs"
+means = np.array([[5, 5], [10, 10]])
+pred = get_labels(X, means)
+```
+
+Benchmarking this function gives us a baseline of **1.26s** on an AMD 3970X CPU.
+
+Compiling this function is now as easy as wrapping it with `torch.compile` and
+executing it with the example inputs
+
+```python
+compiled_fn = torch.compile(get_labels)
+new_pred = compiled_fn(X, means)
+assert np.allclose(prediction, new_pred)
+```
+
+The compiled function yields a 9x speed-up when running it on 1 core. Even
+better, since the compiled code also runs on multiple cores, we get a **57x speed-up**
+when running it on 32 cores. Note that vanilla NumPy always runs on
+just one core.
```

Comment:
Have you compared the performance to other options for compiling numpy. The `get_labels` example contains functions which are supported by `numba`, that might be an interesting datapoint to see how the `torch.compile` speedup compares to `numba.jit`

Reply:
Not really, as this would just make our announcement post way too long. I will expect other people to put up posts comparing this and other approaches (julia / numba / torch.jit / mojo), but that's beyond the scope of this post IMO

Reply:
I plan to write a follow-up, based on https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/168 or https://github.com/Quansight-Labs/numpy_pytorch_interop/tree/main/e2e/smoke, comparing to numba.jit may fit there.

Reply:
That'd be super cool!

Reply:
Weirdly enough, throwing `@numba.njit` in place of `torch.compile` in this example runs into several rough edges of numba:

- `np.linalg.norm(..., axis=2)` does not compile, chokes on `axis`.
-  Replacing `np.linalg.norm` with 

```
def norm(a, axis):
    s = (a.conj() * a).real
    return np.sqrt(s.sum(axis=axis))
```

and `njit`-ting both `norm`  and `get_labels` gives a _slowdown_ of about 60%.

- trying `@njit(parallel=True)` on `get_labels` crashes the compiler, again on the `axis` argument (`TypeError: Failed in nopython mode pipeline (step: Preprocessing for parfors) got an unexpected keyword argument 'axis'`)

- trying `@njit(parallel=True)` on the `norm` only and `@njit` on `get_labels` yields a slowdown w.r.t. numpy of 'only' 20%. 


Got to admit I never had much luck with numba in non-toy situations.